lib/Sema/CSDiagnostics.h

```diff
@@ -701,6 +701,31 @@ class ContextualFailure : public FailureDiagnostic {
 
   static Optional<Diag<Type, Type>>
   getDiagnosticFor(ContextualTypePurpose context, Type contextualType);
+
+protected:
+  bool exprNeedsParensBeforeAddingAs(const Expr *expr) const {
```

Comment:
Let's not duplicate the logic here. You could extract these methods from `MissingExplicitConversionFailure` into a static function in CSDiagnostics.cpp just need to add `DeclContext *` as an argument to both of them.